statsmodels/tsa/statespace/regime_switching/kim_filter.py

```diff
@@ -0,0 +1,635 @@
+    multivariate_normal_logpdf = multivariate_normal.logpdf
+except ImportError:
+    def multivariate_normal_logpdf(x, mean=None, cov=None):
+        n = x.shape[0]
+        if np.linalg.matrix_rank(cov) < n:
+            raise RuntimeError('Cov matrix is singular.')
+        #TODO: to test this
+        return -0.5 * n * np.log(np.pi) - 0.5 * np.log(np.linalg.det(cov)) - \
+                0.5 * (x - mean).T.dot(np.linalg.inv(cov)).dot(x - mean)
+
+class KimFilter(object):
+    '''
+    Kim Filter
+    '''
+
+    def __init__(self, k_endog, k_states, k_regimes, dtype=np.float64,
+            loglikelihood_burn=0, designs=None, obs_intercepts=None,
+            obs_covs=None, transitions=None, state_intercepts=None,
+            selections=None, state_covs=None, regime_switch_probs=None,
+            **kwargs):
+
+        if k_regimes < 1:
+            raise ValueError('Only multiple regimes are available.'
+                    'Consider using regular KalmanFilter.')
+
+        self._k_endog = k_endog
+        self._k_states = k_states
+        self._k_regimes = k_regimes
+        self._dtype = dtype
+        self._loglikelihood_burn = loglikelihood_burn
+
+        self.per_regime_dims = {
+                'designs': 3,
+                'obs_intercepts': 2,
+                'obs_covs': 3,
+                'transitions': 3,
+                'state_intercepts': 2,
+                'selections': 3,
+                'state_covs': 3
+        }
+
+        designs = self._prepare_data_for_regimes(designs,
+                self.per_regime_dims['designs'])
+        obs_intercepts = self._prepare_data_for_regimes(obs_intercepts,
+                self.per_regime_dims['obs_intercepts'])
+        obs_covs = self._prepare_data_for_regimes(obs_covs,
+                self.per_regime_dims['obs_covs'])
+        transitions = self._prepare_data_for_regimes(transitions,
+                self.per_regime_dims['transitions'])
+        state_intercepts = self._prepare_data_for_regimes(state_intercepts,
+                self.per_regime_dims['state_intercepts'])
+        selections = self._prepare_data_for_regimes(selections,
+                self.per_regime_dims['selections'])
+        state_covs = self._prepare_data_for_regimes(state_covs,
+                self.per_regime_dims['state_covs'])
+
+        # Kalman filters for each regime
+
+        kwargs['alternate_timing'] = True
+        self._regime_kalman_filters = [KalmanFilter(k_endog, k_states,
+                dtype=dtype, loglikelihood_burn=loglikelihood_burn,
+                design=designs[i], obs_intercept=obs_intercepts[i],
+                obs_cov=obs_covs[i], transition=transitions[i],
+                state_intercept=state_intercepts[i], selection=selections[i],
+                state_cov=state_covs[i], **kwargs) for i in range(k_regimes)]
+
+        self.set_regime_switch_probs(regime_switch_probs)
+
+    def set_regime_switch_probs(self, regime_switch_probs):
```

Comment:
I think it might be more descriptive to call this matrix `regime_transition`; what do you think?